### DISCUSSION TITLE
Apply multi-qubit gates to an MPS

### DISCUSSION THREAD

**Discussion by vincentmr** [jcmgray/quimb#185] [Q&A] [answered]
Suppose I would like to simulate a quantum circuit as an MPS. This can be done initializing the circuit with 

```python
state = MPS_computational_state("0" * n)
```

and then applying/contracting the gates in `swap+split` mode. I am wondering, what the best way to deal with multi-qubit unitaries, since it appears `swap+split` is expecting one- and two-qubit operators only. Possibly

- generalize `swap+split`
- a method that creates MPOs given a gate and qubits
- use slices (not sure what are the limits of that though, haven't played with this yet)

**Comment by vincentmr** [comment]
For everyone's interest, I implemented some quick & dirty functions that return an MPO [here](https://github.com/PennyLaneAI/pennylane/blob/prototype/quimb/pennylane/devices/mps/apply_operation.py#L85). I used it to solve a VQE problem, so it's not thoroughly tested, but it seems to work.

**Comment by jcmgray** [comment, marked as answer]
I think the best approach, which is not implemented yet sadly, would be as you say,

1. Convert the gate to a MPO acting on covering set of qubits, (I do have the code for this somewhere)
2. Do an application and compress just on this minimal set of qubits

In meantime one could also manually decompose into two body gates.

**Reply by vincentmr** [reply]
As follow up questions: 

- It seems one can apply an MPS to an MPO, but not the other way around. Is that right? Any reason for that? 
- Can `MatrixProductOperator.apply` work with a subset of sites (I'm currently filling-in with identities)?

**Reply by jcmgray** [reply]
* No, I haven't been using and thus updating the 1D tensor network stuff recently so its a little unpolished and some conveniences like that are missing. There are various new methods for applying a MPO to an MPS here - https://github.com/jcmgray/quimb/blob/main/quimb/experimental/tensor_1d_mpo_gate_methods.py
* Not currently, that would be nice functionality, the only trickiness being aligning the partial MPO with the MPS

**Comment by jcmgray** [comment]
Multi-qubit gates are now handled by `CircuitMPS` via an MPO representation. It currently works via the dense representation so things like n-controlled gates which have a direct low-rank MPO representation are not implemented yet.
